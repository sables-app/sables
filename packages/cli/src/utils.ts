import type { SablesBuildOptions } from "./types.js";

/** @public */
export const serverTargets = {
  /**
   * Server build target for the Node.js runtime
   *
   * @public
   */
  NODE: "node",
  /**
   * Server build target for the Cloudflare Workers runtime
   *
   * @public
   */
  WORKER: "worker",
} as const;

/**
 * A utility to retrieve boolean values from the Node.js env.
 * Intended to be used within Sables configuration files.
 *
 * @public
 */
export function envFlag(name: string) {
  return !!JSON.parse(process.env[name] || "false");
}

/**
 * A utility to assist in writing configuration for Sables.
 *
 * @public
 */
export function defineConfig(config: SablesBuildOptions): SablesBuildOptions {
  return config;
}

defineConfig.serverTargets = serverTargets;
