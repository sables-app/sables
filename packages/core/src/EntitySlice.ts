import type * as ReduxToolkit from "@reduxjs/toolkit";

import { createEntityAdapter } from "../deps.js";
import {
  createNotableEntities,
  entityAdapterToReducers,
  EntityReducers,
  NotableEntities,
} from "./Entity.js";
import { createSlice } from "./Slice.js";
import { EnhancedSlice } from "./types.js";

type EntityAdapterOptions<Entity> = Parameters<
  typeof createEntityAdapter<Entity>
>[0];

type EntitySliceState<
  Entity,
  Adjectives extends NotableEntities.Adjectives,
  SingularName extends string
> = ReduxToolkit.EntityState<Entity> &
  NotableEntities.SliceState<Entity, Adjectives, SingularName>;

type EntitySliceReducers<
  Entity,
  Adjectives extends NotableEntities.Adjectives,
  SingularName extends string,
  PluralName extends string
> = NotableEntities.Instance<Entity, Adjectives, SingularName>["reducers"] &
  EntityReducers.Reducers<Entity, SingularName, PluralName>;

type EntitySliceSelectors<
  Entity,
  Adjectives extends NotableEntities.Adjectives,
  SingularName extends string,
  PluralName extends string
> = NotableEntities.Selectors<
  Record<PluralName, EntitySliceState<Entity, Adjectives, SingularName>> &
    Record<string, unknown>,
  Entity,
  Adjectives,
  SingularName
>;
type EntitySliceMixin<
  Entity,
  Adjectives extends NotableEntities.Adjectives,
  SingularName extends string,
  PluralName extends string
> = {
  selectors: EntitySliceSelectors<Entity, Adjectives, SingularName, PluralName>;
};

/**
 * @see {createEntitySlice}
 *
 * @public
 */
export type EntitySlice<
  Entity,
  Adjectives extends NotableEntities.Adjectives,
  SingularName extends string,
  PluralName extends string
> = EnhancedSlice<
  ReduxToolkit.Slice<
    EntitySliceState<Entity, Adjectives, SingularName>,
    EntitySliceReducers<Entity, Adjectives, SingularName, PluralName>,
    PluralName
  >
> &
  EntitySliceMixin<Entity, Adjectives, SingularName, PluralName>;

/**
 * Creates a slice to store a specific entity.
 *
 * The slice is created by composing the following functions together:
 *
 * - `createEntityAdapter`
 * - `entityAdapterToReducers`
 * - `createNotableEntities`
 * - `createSlice`
 * - `entityAdapter.getSelectors`
 * - `notableEntities.getSelectors`
 *
 * @example
 *
 * type Book = { id: string; name: string };
 *
 * const booksSlice = createEntitySlice<Book>().setReducer(
 *   "book",
 *   "books",
 *   {
 *     best: (id: string) => id,
 *   }
 * );
 *
 * const { addBook, denoteBestBook } = booksSlice.actions;
 * const { selectBestBook } = booksSlice.selectors;
 *
 * store.dispatch(
 *   addBook({
 *     id: "1",
 *     name: "Goosebumps: Night of the Living Dummy",
 *   })
 * );
 *
 * store.dispatch(denoteBestBook("1"));
 *
 * // Returns: { id: "1", name: "Goosebumps: Night of the Living Dummy" }
 * selectBestBook(store.getState());
 *
 * @see {createEntityAdapter}
 * @see {entityAdapterToReducers}
 * @see {createNotableEntities}
 * @see {createSlice}
 *
 * @public
 */
export function createEntitySlice<Entity>(
  adapterOptions?: EntityAdapterOptions<Entity>
) {
  return {
    setReducer<
      Adjectives extends NotableEntities.Adjectives | void,
      SingularName extends string,
      PluralName extends string = `${SingularName}s`
    >(
      singularName: SingularName,
      plural: PluralName,
      adjectives?: Adjectives
    ): EntitySlice<
      Entity,
      Adjectives extends NotableEntities.Adjectives
        ? Adjectives
        : // eslint-disable-next-line @typescript-eslint/ban-types
          {},
      SingularName,
      PluralName
    > {
      const pluralName = plural ?? (`${singularName}s` as const);
      const entityAdapter = createEntityAdapter<Entity>(adapterOptions);
      const adapterReducers = entityAdapterToReducers(
        entityAdapter,
        singularName,
        pluralName
      );

      const notableEntities = createNotableEntities(
        entityAdapter,
        singularName,
        adjectives || {}
      );
      const initialState = {
        ...entityAdapter.getInitialState(),
        ...notableEntities.getInitialState(),
      };

      const slice = createSlice(pluralName, initialState).setReducer(
        (builder) =>
          builder.addCases({ ...adapterReducers, ...notableEntities.reducers })
      );

      const selectors = {
        ...entityAdapter.getSelectors(slice.selector),
        ...notableEntities.getSelectors(slice.selector),
      };

      (slice as any).selectors = selectors;

      return slice as any;
    },
  };
}
