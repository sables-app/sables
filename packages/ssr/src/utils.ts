import { SSR_ATTRIBUTE, ssrAttrValues } from "@sables/framework";

export function shouldHydrate() {
  return (
    typeof document == "object" &&
    !!document.querySelector(
      `[${SSR_ATTRIBUTE}="${ssrAttrValues.SHOULD_HYDRATE}"]`
    )
  );
}
