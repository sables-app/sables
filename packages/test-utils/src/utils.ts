import { VITest } from "./types.js";

type LogKey = "info" | "log" | "warn" | "error";

export function silenceLogs(vitest: VITest, key: LogKey) {
  const { afterAll, beforeAll, vi } = vitest;
  let originalMethod: typeof console[LogKey];

  beforeAll(() => {
    originalMethod = console[key];
    console[key] = vi.fn();
  });

  afterAll(() => {
    console[key] = originalMethod;
  });
}
