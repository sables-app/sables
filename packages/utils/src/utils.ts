import type * as ReduxToolkit from "@reduxjs/toolkit";
import type * as Redux from "redux";

import { getBaseURL } from "./env.js";
import type {
  MutableReferenceObject,
  SlicesToReducersMapObject,
} from "./types.js";

/** @internal */
export function createMutableRef<T>(errorMessage?: string, initialValue?: T) {
  const ref: MutableReferenceObject<T> = {
    current: initialValue,
    error() {
      return new ReferenceError(
        errorMessage || "Value has not been initialized.",
      );
    },
    demand() {
      if (ref.current === undefined) {
        throw ref.error();
      }

      return ref.current;
    },
  };

  return ref;
}

/** @internal */
export function demandValue<T>(
  value?: T,
  message = "An expected value was not defined!",
) {
  return createMutableRef<T>(message, value).demand();
}

/** @internal */
export function composeMiddleware(
  ...middleware: Redux.Middleware[]
): Redux.Middleware {
  type InitFn = ReturnType<Redux.Middleware>;

  function identity<T>(val: T): T {
    return val;
  }

  function composeInitFns(funcs: InitFn[], next: Redux.Dispatch) {
    const last = funcs[funcs.length - 1];
    const rest = funcs.slice(0, -1);

    return rest.reduceRight(
      (composed, initFn) => initFn(composed),
      last(next || identity),
    );
  }

  return (middlewareAPI) => {
    const chain = middleware.map((middleware) => middleware(middlewareAPI));
    return (next) => {
      const dispatch = composeInitFns(chain, next);
      return (action) => dispatch(action);
    };
  };
}

/** @internal */
export function resolveAssetPath(assetPath: string) {
  const assetsBase = getBaseURL();

  try {
    return new URL(assetPath, assetsBase).toString();
  } catch (error) {
    return new URL(
      assetPath,
      new URL(assetsBase, "http://example.com/").toString(),
    ).pathname;
  }
}

/** @internal */
export function createReducersMapObject<P extends [...ReduxToolkit.Slice[]]>(
  ...slices: P
): SlicesToReducersMapObject<P> {
  return slices.reduce<any>((result, slice) => {
    return { ...result, [slice.name]: slice.reducer };
  }, {});
}

/** @internal */
export function capitalize<T extends string>(name: T) {
  return `${name.charAt(0).toUpperCase()}${name.slice(1)}` as Capitalize<T>;
}

/**
 * CSS-in-JS utility to facilitate in creating responsive design tokens
 *
 * @public
 */
export function responsiveToken(scaling: number[], base: number, unit = "rem") {
  return scaling.map((scale) => `${scale * base}${unit}`);
}
