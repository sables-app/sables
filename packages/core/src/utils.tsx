import { createMutableRef, MutableReferenceObjectValue } from "@sables/utils";

import type * as ReduxToolkit from "@reduxjs/toolkit";
import type { ComponentProps, ElementType, FunctionComponent } from "react";
import type * as Redux from "redux";

import {
  LIFECYCLE_REF_MESSAGE,
  ROUTES_COLLECTION_REF_MESSAGE,
  SYMBOL_EFFECT_API_LIFECYCLE,
  SYMBOL_EFFECT_API_ROUTES,
  SYMBOL_LAZY_META,
  SYMBOL_LAZY_SELECTOR,
} from "./constants.js";
import { ActionSubject } from "./main.js";
import type {
  AnyLazySelector,
  DefaultEffectAPI,
  LazyMeta,
  ObjectWithLazyMeta,
} from "./types.js";

/** @internal */
export function hasLazyMeta<T extends object>(
  value: T,
): value is T & ObjectWithLazyMeta {
  return Object.hasOwn(value, SYMBOL_LAZY_META);
}

/** @internal */
export function createLazyMeta(): LazyMeta {
  return {
    observers: new Set(),
    slices: new Set(),
  };
}

/** @internal */
export function isEnhancedSelector(value: unknown): value is AnyLazySelector {
  return (
    typeof value == "function" && Object.hasOwn(value, SYMBOL_LAZY_SELECTOR)
  );
}

/** @internal */
export function getSlicesFromSelector(
  selector: ReduxToolkit.Selector,
): ReduxToolkit.Slice[] {
  return isEnhancedSelector(selector)
    ? [...selector[SYMBOL_LAZY_META].slices]
    : [];
}

type LifecycleRef<
  StoreState extends Record<string, unknown> = Record<string, unknown>,
> = DefaultEffectAPI<StoreState>[typeof SYMBOL_EFFECT_API_LIFECYCLE];

type RoutesCollectionRef<
  StoreState extends Record<string, unknown> = Record<string, unknown>,
> = DefaultEffectAPI<StoreState>[typeof SYMBOL_EFFECT_API_ROUTES];

/** @internal */
export function createEffectAPIDefaults<
  StoreState extends Record<string, unknown> = Record<string, unknown>,
>({
  actions$,
  dispatch,
  getState,
  lifecycleRef,
  routesCollectionRef,
}: {
  actions$?: ActionSubject;
  dispatch: ReduxToolkit.Dispatch;
  getState: Redux.Store<StoreState>["getState"];
  lifecycleRef?: LifecycleRef<StoreState>;
  routesCollectionRef?: RoutesCollectionRef<StoreState>;
}): DefaultEffectAPI<StoreState> {
  return {
    actions$: actions$ || new ActionSubject(),
    dispatch,
    getState,
    [SYMBOL_EFFECT_API_LIFECYCLE]:
      lifecycleRef ||
      createMutableRef<MutableReferenceObjectValue<LifecycleRef>>(
        LIFECYCLE_REF_MESSAGE,
      ),
    [SYMBOL_EFFECT_API_ROUTES]:
      routesCollectionRef ||
      createMutableRef<MutableReferenceObjectValue<RoutesCollectionRef>>(
        ROUTES_COLLECTION_REF_MESSAGE,
      ),
  };
}

type WithPropsTransformHook<Props> = (props: Props) => Props;
type WithPropsInput<Props> = Props | WithPropsTransformHook<Props>;

/**
 * A higher-order component that assigns props to the given component.
 *
 * @see {@link https://sables.dev/api#withprops withProps documentation}
 *
 * @example
 *
 * const Link = withProps("a", { className: "link" });
 * const ButtonLink = withProps(Link, { role: "button" });
 * const A11yButtonLink = withProps(
 *   ButtonLink,
 *   (props) => ({ "aria-label": props.title })
 * );
 *
 * @public
 */
export function withProps<C extends ElementType>(
  Component: C,
  input: WithPropsInput<ComponentProps<C>>,
): FunctionComponent<ComponentProps<C>> {
  type Props = ComponentProps<C>;

  function isWithPropsTransformHook(
    value: WithPropsInput<Props>,
  ): value is WithPropsTransformHook<Props> {
    return typeof value == "function";
  }

  const useProps = isWithPropsTransformHook(input)
    ? input
    : (props: Props) => ({ ...input, ...props });

  function ComponentWithProps(props: Props) {
    return <Component {...useProps(props)} />;
  }

  const originalDisplayName =
    (typeof Component == "function" && Component.displayName) || "Component";

  ComponentWithProps.displayName = `withProps(${originalDisplayName})`;

  return ComponentWithProps;
}
