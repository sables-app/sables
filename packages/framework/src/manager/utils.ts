import {
  DefaultEffectAPI,
  SYMBOL_EFFECT_API_LIFECYCLE,
  SYMBOL_EFFECT_API_ROUTES,
  SYMBOL_MANAGER_EFFECT_API,
} from "@sables/core";
import { isSSREnv } from "@sables/utils";

import { LifecycleState, ServerRequestStateRef } from "../LifecycleContext.js";
import { ConfigureManagerFn, ConfigureManagerParams, Manager } from "../types.js";
import { createManager } from "./Manager.js";

export * from "@sables/utils";

function assertManagerConsistency<
  EffectAPI extends DefaultEffectAPI = DefaultEffectAPI
>(manager: Manager<any, EffectAPI>) {
  const effectAPIRefOnManager = manager[SYMBOL_MANAGER_EFFECT_API];
  const lifecycleRefOnManager = manager[SYMBOL_EFFECT_API_LIFECYCLE];
  const routeCollectionRefOnManager = manager[SYMBOL_EFFECT_API_ROUTES];

  const effectAPI = effectAPIRefOnManager.demand();

  const lifecycleRefOnEffectAPI = effectAPI[SYMBOL_EFFECT_API_LIFECYCLE];
  const routeCollectionRefOnEffectAPI = effectAPI[SYMBOL_EFFECT_API_ROUTES];

  routeCollectionRefOnManager.demand();
  lifecycleRefOnManager.demand();
  lifecycleRefOnEffectAPI.demand();

  if (lifecycleRefOnManager.current !== lifecycleRefOnEffectAPI.current) {
    throw new Error("Inconsistent lifecycle state detected.");
  }

  if (
    routeCollectionRefOnEffectAPI.current &&
    routeCollectionRefOnEffectAPI.current !==
      routeCollectionRefOnManager.current
  ) {
    throw new Error(
      [
        "Multiple route collections detected.",
        "When a route collection object is defined on the effect API, it must match the one defined on the manager.",
      ].join("\n")
    );
  }
}

const DEFAULT_PARAMS: ConfigureManagerParams = {};

function getConfigureManagerParams(
  serverRequestStateRef: ServerRequestStateRef
): ConfigureManagerParams {
  if (!isSSREnv()) {
    return DEFAULT_PARAMS;
  }

  const { href } = serverRequestStateRef.demand();

  return {
    initialLocation: href,
  };
}

/** @internal */
export function resolveConfigureManager<M extends Manager<any, any>>(
  lifecycleState: LifecycleState,
  configureManager: ConfigureManagerFn<M> | undefined
): M {
  const { managerRef, serverRequestStateRef } = lifecycleState;

  if (!managerRef.current) {
    const managerCreator = configureManager || createManager;
    managerRef.current = managerCreator(
      getConfigureManagerParams(serverRequestStateRef)
    );
    managerRef.current[SYMBOL_EFFECT_API_LIFECYCLE].current = lifecycleState;
    assertManagerConsistency(managerRef.current);
  }

  return managerRef.current as M;
}

/**
 * A utility function to assist in defining `configureManager` functions.
 *
 * @example
 *
 * export const configureManager = defineConfigureManager(
 *   ({ initialLocation }) => {
 *     return createManager({ initialLocation });
 *   }
 * );
 *
 * @public
 */
export function defineConfigureManager<M extends Manager<any, any>>(
  configureManager: ConfigureManagerFn<M>
) {
  return configureManager;
}
