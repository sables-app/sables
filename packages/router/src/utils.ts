import {
  DefaultEffectAPI,
  SYMBOL_EFFECT_API_LIFECYCLE,
  SYMBOL_EFFECT_API_ROUTES,
} from "@sables/core";
import { isSSREnv } from "@sables/utils";

import type * as History from "history";
import { createPath } from "history";

import { ensureLocation, pushLocation, replaceLocation } from "./actions.js";
import { AnyRouteReference } from "./Routes.js";
import { isRoutesCollection, RoutesCollection } from "./RoutesCollection.js";
import type {
  BuildHrefParams,
  BuildLinkParams,
  LocationChangeAction,
  RegisterDynamicImportFn,
  RouteHref,
  RouteLink,
} from "./types.js";

/**
 * Used to compare location changes to determine whether
 * a route transition should start.
 * Only the pathname and search are compared. Hashes are not
 * considered to be a part of a route path.
 * @internal
 */
export function areLocationChangesRouterEquivalent(
  changeA?: LocationChangeAction["payload"],
  changeB?: LocationChangeAction["payload"]
): boolean {
  return !!(
    changeA &&
    changeB &&
    changeA.location.pathname === changeB.location.pathname &&
    changeA.location.search === changeB.location.search
  );
}

/** @internal */
export function isPartialHistoryPath(
  value?: AnyRouteReference | Partial<History.Path> | BuildHrefParams
): value is Partial<History.Path> {
  return (
    typeof value === "object" &&
    !Array.isArray(value) &&
    typeof (value as any).id === "undefined"
  );
}

/**
 * `import('history').createPath()` also returns the root
 * path when the given values are undefined.
 */
const FALLBACK_HREF = "/";

/**
 * Builds a hyperlink for a given route.
 *
 * @public
 */
export function buildHref(...args: BuildHrefParams): RouteHref {
  const [route, params, opts] = args;

  if (isPartialHistoryPath(route)) {
    return createPath(route) as RouteHref;
  }

  if (!route) return FALLBACK_HREF;

  return route.build(params, opts);
}

/**
 * Builds an object to interact with a route.
 *
 * @see {RouteLink}
 *
 * @public
 */
export function buildLink(...args: BuildLinkParams): RouteLink {
  const [dispatch, buildHrefArgs] = args;
  const href = buildHref(...buildHrefArgs);

  return {
    href,
    ensureLocation() {
      dispatch(ensureLocation(href));
    },
    pushLocation() {
      dispatch(pushLocation(href));
    },
    replaceLocation() {
      dispatch(replaceLocation(href));
    },
  };
}

/** @internal */
export function demandRoutesCollectionFromEffectAPI<
  EffectAPI extends DefaultEffectAPI = DefaultEffectAPI
>(effectAPI: EffectAPI): RoutesCollection<EffectAPI> {
  const routesCollectionRef = effectAPI[SYMBOL_EFFECT_API_ROUTES];

  if (!isRoutesCollection<EffectAPI>(routesCollectionRef.current)) {
    throw new Error("Routes must be made available to use this effect.");
  }

  return routesCollectionRef.current;
}

/** @internal */
export function createDynamicImportRegistrar<
  EffectAPI extends DefaultEffectAPI = DefaultEffectAPI
>(effectAPI: EffectAPI): RegisterDynamicImportFn {
  return function registerDynamicImport(dynamicImport) {
    if (!isSSREnv()) return;

    effectAPI[SYMBOL_EFFECT_API_LIFECYCLE].demand()
      .serverRequestStateRef.demand()
      .invokedImporters.add(String(dynamicImport));
  };
}
