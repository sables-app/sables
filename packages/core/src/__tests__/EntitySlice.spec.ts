import * as vitest from "vitest";
import { assertType, describe, expect, it, test } from "vitest";

import { createEntitySlice } from "../EntitySlice.js";
import { createTestStore } from "./utils.js";

describe("EntitySlice", () => {
  type Book = { id: string; name: string };

  describe("createEntitySlice", () => {
    it("creates a slice with entity logic", async () => {
      const booksSlice = createEntitySlice<Book>().setReducer("book", "books", {
        best: (id: string) => id,
      });

      // @ts-expect-error The action doesn't accept invalid values
      booksSlice.actions.addBook(true);

      // Actions from Redux Toolkit should exist
      booksSlice.actions.addMany([]);

      const { store } = createTestStore(vitest);

      expect(store.getState()).toEqual(undefined);

      const addBookAction = booksSlice.actions.addBook({
        id: "1",
        name: "Goosebumps: Night of the Living Dummy",
      });

      assertType<{
        payload: Book;
        type: "books/addBook";
      }>(addBookAction);

      expect(addBookAction).toMatchSnapshot();

      store.dispatch(addBookAction);

      expect(store.getState()).toMatchSnapshot();

      const { selectBestBook } = booksSlice.selectors;

      // @ts-expect-error The selector shouldn't exist
      booksSlice.selectors.selectWorstBook;
      // The selector should exist
      assertType<(state: Record<string, unknown>) => Book[]>(
        booksSlice.selectors.selectAllBooks
      );

      expect(selectBestBook(store.getState())).toEqual(undefined);

      store.dispatch(booksSlice.actions.denoteBestBook("1"));

      expect(selectBestBook(store.getState())).toMatchSnapshot();
    });

    test("optional adjectives", async () => {
      const booksSlice = createEntitySlice<Book>().setReducer("book", "books");

      // @ts-expect-error The selector shouldn't exist
      const { selectBestBook } = booksSlice.selectors;

      expect(selectBestBook).toEqual(undefined);

      // @ts-expect-error The selector shouldn't exist
      expect(booksSlice.selectors.selectWorstBook).toEqual(undefined);
    });

    test("entity interface", async () => {
      interface Foo1 {
        bar: string;
      }
      type Foo2 = {
        bar: string;
      };

      // @ts-expect-error Types created with `interface` aren't compatible
      // with `Record<string, unknown>` by default.
      assertType<Record<string, unknown>>({} as Foo1);

      // Types created with `type` are compatible
      // with `Record<string, unknown>` by default.
      assertType<Record<string, unknown>>({} as Foo2);

      // No type error should occur when a entity type created with `interface` is used.
      createEntitySlice<Foo1>().setReducer("foo", "foos");

      // No type error should occur when a entity type created with `type` is used.
      createEntitySlice<Foo2>().setReducer("foo", "foos");
    });
  });
});
