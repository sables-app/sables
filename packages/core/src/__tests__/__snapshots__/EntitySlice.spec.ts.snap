// Vitest Snapshot v1

exports[`EntitySlice > createEntitySlice > creates a slice with entity logic 1`] = `
{
  "error": false,
  "meta": {
    Symbol(@sables/lazyMeta): {
      "observers": Set {},
      "slices": Set {
        {
          "actions": {
            "addBook": [Function],
            "addBooks": [Function],
            "addMany": [Function],
            "addOne": [Function],
            "denoteBestBook": [Function],
            "removeAll": [Function],
            "removeAllBooks": [Function],
            "removeBook": [Function],
            "removeBooks": [Function],
            "removeMany": [Function],
            "removeOne": [Function],
            "setAll": [Function],
            "setAllBooks": [Function],
            "setBook": [Function],
            "setBooks": [Function],
            "setMany": [Function],
            "setOne": [Function],
            "updateBook": [Function],
            "updateBooks": [Function],
            "updateMany": [Function],
            "updateOne": [Function],
            "upsertBook": [Function],
            "upsertBooks": [Function],
            "upsertMany": [Function],
            "upsertOne": [Function],
          },
          "caseReducers": {
            "addBook": [Function],
            "addBooks": [Function],
            "addMany": [Function],
            "addOne": [Function],
            "denoteBestBook": [Function],
            "removeAll": [Function],
            "removeAllBooks": [Function],
            "removeBook": [Function],
            "removeBooks": [Function],
            "removeMany": [Function],
            "removeOne": [Function],
            "setAll": [Function],
            "setAllBooks": [Function],
            "setBook": [Function],
            "setBooks": [Function],
            "setMany": [Function],
            "setOne": [Function],
            "updateBook": [Function],
            "updateBooks": [Function],
            "updateMany": [Function],
            "updateOne": [Function],
            "upsertBook": [Function],
            "upsertBooks": [Function],
            "upsertMany": [Function],
            "upsertOne": [Function],
          },
          "entityAdapter": {
            "addMany": [Function],
            "addOne": [Function],
            "getInitialState": [Function],
            "getSelectors": [Function],
            "removeAll": [Function],
            "removeMany": [Function],
            "removeOne": [Function],
            "selectId": [Function],
            "setAll": [Function],
            "setMany": [Function],
            "setOne": [Function],
            "sortComparer": false,
            "updateMany": [Function],
            "updateOne": [Function],
            "upsertMany": [Function],
            "upsertOne": [Function],
          },
          "getInitialState": [Function],
          "name": "books",
          "reducer": [Function],
          "selectBooksState": [Function],
          "selector": [Function],
          "selectors": {
            "selectAll": [Function],
            "selectAllBooks": [Function],
            "selectBestBook": [Function],
            "selectBestBookPayload": [Function],
            "selectBookById": [Function],
            "selectBookCount": [Function],
            "selectBookDictionary": [Function],
            "selectBookIds": [Function],
            "selectById": [Function],
            "selectEntities": [Function],
            "selectIds": [Function],
            "selectTotal": [Function],
          },
        },
      },
    },
  },
  "payload": {
    "id": "1",
    "name": "Goosebumps: Night of the Living Dummy",
  },
  "type": "books/addBook",
}
`;

exports[`EntitySlice > createEntitySlice > creates a slice with entity logic 2`] = `
{
  "books": {
    "bestBookPayload": undefined,
    "entities": {
      "1": {
        "id": "1",
        "name": "Goosebumps: Night of the Living Dummy",
      },
    },
    "ids": [
      "1",
    ],
  },
}
`;

exports[`EntitySlice > createEntitySlice > creates a slice with entity logic 3`] = `
{
  "id": "1",
  "name": "Goosebumps: Night of the Living Dummy",
}
`;
